independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,143 @@
+    private final AtomicInteger referenceCounter = new AtomicInteger(2);
+
+    private JarFileReference(AtomicReference<JarFileReference> jarFileReference, JarFile jarFile) {
+        this.jarFileReference = jarFileReference;
+        this.jarFile = jarFile;
+    }
+
+    /**
+     * Increase the readers counter of the jarFile.
+     *
+     * @return true if the acquire succeeded: it's now safe to access and use the inner jarFile.
+     *         false if the jar reference is going to be closed and then no longer usable.
+     */
+    private boolean acquire() {
+        while (true) {
+            int count = referenceCounter.get();
+            if (count == 0) {
+                return false;
+            }
+            if (referenceCounter.compareAndSet(count, count + 1)) {
+                return true;
+            }
+        }
+    }
+
+    /**
+     * Decrease the readers counter of the jarFile.
+     * If the counter drops to 0 and a release has been requested also closes the jarFile.
+     *
+     * @return true if the release also closes the underlying jarFile.
+     */
+    private boolean release() {
+        while (true) {
+            int count = referenceCounter.get();
+            if (count <= 0) {
+                throw new IllegalStateException(
+                        "The reference counter cannot be negative, found: " + (referenceCounter.get() - 1));
+            }
+            if (referenceCounter.compareAndSet(count, count - 1)) {
+                if (count == 1) {
+                    try {
+                        jarFile.close();
+                    } catch (IOException e) {
+                        // ignore
+                    } finally {
+                        jarFileReference.compareAndSet(this, null);
+                    }
+                    return true;
+                }
+                return false;
+            }
+        }
+    }
+
+    /**
+     * Ask to close this reference.
+     * If there are no readers currently accessing the jarFile also close it, otherwise defer the closing when the last reader
+     * will leave.
+     */
+    void close() {
+        release();
+    }
+
+    @FunctionalInterface
+    interface JarFileConsumer<T> {
+        T apply(JarFile jarFile, Path jarPath, String resource);
+    }
+
+    static <T> T withJarFile(AtomicReference<JarFileReference> jarFileReference, Path jarPath, String resource,
+            JarFileConsumer<T> fileConsumer) {
+
+        // Happy path: the jar reference already exists and it's ready to be used
+        final var localJarFileRef = jarFileReference.get();
+        if (localJarFileRef != null && localJarFileRef.acquire()) {
+            return consumeSharedJarFile(jarPath, resource, fileConsumer, localJarFileRef);
+        }
+
+        // There's no valid jar reference, so load a new one
+        final var newJarFileRef = JarFileReference.loadAcquiredJarFile(jarFileReference, jarPath);
```

Comment:
I might be wrong, but this seems racy here?
In case we're not on the happy path, it would seem two(or more) threads could end up triggering loading of a new `JarFile`.
This wouldn't be coooperative, and more importantly I don't see where the additional references get closed.

I'd argue that this is actually a likely path, as the opening of such Jar is an high latency operation therefore there's good chances for this line to be hit concurrently.

Reply:
ah my bad it gets closed right after.. still I'm wondering if it could be more co-operative?

Reply:
@mariofusco / @franz1981 did you notice this comment here?

I'm not sure but it *could* explain the worsening figures, as I'm quite sure that this pattern would lead concurrent threads needing the jar at about the same jar to open the same jar multiple times.